book/src/dev/rfcs/0003-state-updates.md

```diff
@@ -0,0 +1,305 @@
+# State Updates
+
+- Feature Name: state_updates
+- Start Date: 2020-08-14
+- Design PR: XXX
+- Zebra Issue: XXX
+
+# Summary
+[summary]: #summary
+
+Zebra manages chain state in the `zebra-state` crate, which allows state
+queries via asynchronous RPC (in the form of a Tower service). The state
+system is responsible for contextual verification in the sense of [RFC2],
+checking that new blocks are consistent with the existing chain state before
+committing them. This RFC describes how the state is represented internally,
+and how state updates are performed.
+
+[RFC2]: ./0002-parallel-verification.md
+
+# Motivation
+[motivation]: #motivation
+
+We need to be able to access and modify the chain state, and we want to have
+a description of how this happens and what guarantees are provided by the
+state service.
+
+# Definitions
+[definitions]: #definitions
+
+* **state data**: Any data the state service uses to represent chain state.
+
+* **structural/semantic/contextual verification**: as defined in [RFC2].
+
+* **block chain**: A sequence of valid blocks linked by inclusion of the
+  previous block hash in the subsequent block. Chains are rooted at the
+  *genesis* block and extend to a *tip*.
+
+* **chain state**: The state of the ledger after application of a particular
+  sequence of blocks (state transitions).
+
+* **difficulty**: The cumulative proof-of-work from genesis to the chain tip.
+
+* **best chain**: The chain with the greatest difficulty. This chain
+  represents the consensus state of the Zcash network and transactions.
+
+* **side chain**: A chain which is not contained in the best chain.
+
+* **chain reorganization**: Occurs when a new best chain is found and the
+  previous best chain becomes a side chain.
+
+* **orphaned block**: A block which is no longer included in the best chain.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+XXX fill in after writing other details
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## State Components
+
+Zcash (as implemented by `zcashd`) differs from Bitcoin in its treatment of
+transaction finality. If a new best chain is detected that does not extend
+the previous best chain, blocks at the end of the previous best chain become
+orphaned (no longer included in the best chain). Their state updates are
+therefore no longer included in the best chain's chain state. The process of
+rolling back orphaned blocks and applying new blocks is called a chain
+reorganization. Bitcoin allows chain reorganizations of arbitrary depth,
+while `zcashd` limits reorganizations to 100 blocks.
+
+This difference means that in Bitcoin, chain state only has probabilistic
+finality, while in Zcash, chain state is final once it is beyond the reorg
+limit. To simplify our implementation, we split the representation of the
```

Comment:
I don't expect that this will be a problem, but if it is, we would probably be able to handle a longer reorg with manual intervention by wiping the sled cache and starting over.  This requires manual intervention, but a scenario with 100-block reorgs is a bad one, so I think manual intervention is OK.